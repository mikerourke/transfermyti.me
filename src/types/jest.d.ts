// noinspection JSAnnotator

/**
 * Custom matchers from @testing-library/jest-dom.
 * @see https://github.com/testing-library/jest-dom
 */
declare namespace jest {
  // prettier-ignore
  interface Matchers<R> {
    toBeDisabled(): R;
    toBeEnabled(): R;
    toBeEmptyDOMElement(): R;
    toBeInTheDocument(): R;
    toBeInvalid(): R;
    toBeRequired(): R;
    toBeValid(): R;
    toBeVisible(): R;
    toContainElement(element: HTMLElement | SVGElement | null): R;
    toContainHTML(htmlText: string): R;
    toHaveAccessibleDescription(expectedAccessibleDescription?: string | RegExp): R;
    toHaveAccessibleName(expectedAccessibleName?: string | RegExp): R;
    toHaveAttribute(attr: string, value?: any): R;
    toHaveClass(classNames: string[], options?: { exact: boolean }): R;
    toHaveClass(...classNames: string[]): R;
    toHaveFocus(): R;
    toHaveFormValues(expectedValues: { [name: string]: any }): R;
    toHaveStyle(css: string | Record<string, unknown>): R;
    toHaveTextContent(text: string | RegExp, options?: { normalizeWhitespace: boolean }): R;
    toHaveValue(value: string | string[] | number): R;
    toHaveDisplayValue(value: string | RegExp | (string | RegExp)[]): R;
    toBeChecked(): R;
    toBePartiallyChecked(): R;
    toHaveErrorMessage(text: string | RegExp): R;
  }
}
