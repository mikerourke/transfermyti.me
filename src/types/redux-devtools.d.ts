type ReduxDevToolsReduxAction = any;

type ReduxDevToolsReduxState = any;

/**
 * Allows you to specify if features should be enabled or disabled. If not
 * specified, all the features are enabled. When set as an object, only
 * those included as true will be allowed. Note that except true/false, import
 * and export can be set as "custom" (which is by default for Redux enhancer),
 * meaning that the importing/exporting occurs on the client side.
 * Otherwise, you'll get/set the data right from the monitor part.
 * @see https://github.com/zalmoxisus/redux-devtools-extension/blob/master/docs/API/Arguments.md#features
 */
interface ReduxDevToolsFeaturesPropertyOptions {
  /** Start/pause recording of dispatched actions. */
  pause?: boolean;

  /** Lock/unlock dispatching actions and side effects. */
  lock?: boolean;

  /** Persist states on page reloading. */
  persist?: boolean;

  /** Export history of actions in a file. */
  export?: boolean;

  /** Import history of actions from a file. */
  import?: boolean | "custom";

  /** Jump back and forth (time travelling). */
  jump?: boolean;

  /** Skip (cancel) actions. */
  skip?: boolean;

  /** Drag and drop actions in the history list. */
  reorder?: boolean;

  /** Dispatch custom actions or action creators. */
  dispatch?: boolean;

  /** Generate tests for the selected actions. */
  test?: boolean;
}

/**
 * Represents the object, which contains keys for which to configure the
 * serialization behavior in Redux DevTools. For each of them you can indicate
 * if to include (by setting as true). For function key you can also specify a
 * custom function which handles serialization. See {@link https://github.com/kolodny/jsan} for more details.
 * @example
 *   const store = createStore(reducer, window.__REDUX_DEVTOOLS_EXTENSION__({
 *     serialize: {
 *       options: {
 *        undefined: true,
 *        function: function(fn) { return fn.toString() }
 *       }
 *     }
 *   }
 */
declare interface ReduxDevToolsSerializeOptionsPropertyOptions {
  date?: boolean;
  regex?: boolean;
  undefined?: boolean;
  nan?: boolean;
  infinity?: boolean;
  error?: boolean;
  symbol?: boolean;
  map?: boolean;
  set?: boolean;
  function?: boolean | ((fn: (...args: any[]) => void) => string);
}

/**
 * Options for the `serialize` option for Redux DevTools.
 * @see https://github.com/zalmoxisus/redux-devtools-extension/blob/master/docs/API/Arguments.md#serialize
 */
declare interface ReduxDevToolsSerializeOptions {
  /**
   * Options for serializing types based on the setting.
   * ┌────────────┬──────────────────────────────────────────────────────┐
   * │ Option     │ Description                                          │
   * ├────────────┼──────────────────────────────────────────────────────┤
   * │ undefined  │ Will use regular JSON.stringify to send data         │
   * │            │ (it's the fast mode).                                │
   * ├────────────┼──────────────────────────────────────────────────────┤
   * │ false      │ Will handle also circular references.                │
   * ├────────────┼──────────────────────────────────────────────────────┤
   * │ true       │ Will handle also date, regex, undefined, primitives, │
   * │            │ error objects, symbols, maps, sets and functions.    │
   * ├────────────┼──────────────────────────────────────────────────────┤
   * │ object     │ See `ReduxDevToolsSerializeOptionsPropertyOptions`   │
   * └────────────┴──────────────────────────────────────────────────────┘
   */
  options?: boolean | ReduxDevToolsSerializeOptionsPropertyOptions | undefined;

  /**
   * JSON replacer function used for both actions and states stringify.
   * @see https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#The_replacer_parameter
   */
  replacer?: (key: string, value: unknown) => string;

  /**
   * JSON reviver function used for parsing the imported actions and states.
   * @see https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/JSON/parse#Using_the_reviver_parameter
   */
  reviver?: (key: string, value: unknown) => Record<string, unknown>;

  /**
   * Automatically serialize/deserialize ImmutableJS via remotedev-serialize.
   * Note: Only use this if using the ImmutableJS library.
   */
  immutable?: any;

  /**
   * ImmutableJS Record classes used to make possible restore its instances back
   * when importing, persisting.
   * Note: Only use this if using the ImmutableJS library.
   */
  refs?: any[];
}

/**
 * Options to pass to the Redux DevTools compose function.
 * @see https://github.com/zalmoxisus/redux-devtools-extension/blob/master/docs/API/Arguments.md
 */
declare interface ReduxDevToolsComposeOptions {
  /**
   * The instance name to be shown on the monitor page. Default value is
   * document.title. If not specified and there's no document title, it will
   * consist of tabId and instanceId.
   */
  name?: string;

  /**
   * Action creators functions to be available in the Dispatcher.
   */
  actionCreators?: any | any[];

  /**
   * If more than one action is dispatched in the indicated interval, all new
   * actions will be collected and sent at once. It is the joint between
   * performance and speed. When set to 0, all actions will be sent instantly.
   * Set it to a higher value when experiencing perf issues (also maxAge to
   * a lower value).
   * @default 500
   */
  latency?: number;

  /**
   * Maximum allowed actions to be stored in the history tree. The oldest
   * actions are removed once maxAge is reached. It's critical for performance.
   * @default 50
   */
  maxAge?: number;

  /**
   * If set to true, will include stack trace for every dispatched action,
   * so you can see it in trace tab jumping directly to that part of code.
   * You can use a function (with action object as argument) which should
   * return new Error().stack string, getting the stack outside of reducers.
   * @default false
   */
  trace?: boolean | (() => void);

  /**
   * Maximum stack trace frames to be stored (in case trace option was provided
   * as true). By default, it's 10. Note that, because extension's calls are
   * excluded, the resulted frames could be 1 less. If trace option is a
   * function, traceLimit will have no effect, as it's supposed to be handled
   * there.
   */
  traceLimit?: number;

  /**
   * Options for serialization/deserialization.
   */
  serialize?: ReduxDevToolsSerializeOptions;

  /**
   * Function which takes action object as argument, and should return action
   * object back.
   */
  actionSanitizer?: (
    action: ReduxDevToolsReduxAction,
  ) => ReduxDevToolsReduxAction;

  /**
   * Function which takes state object as argument, and should return state
   * object back.
   */
  stateSanitizer?: (state: ReduxDevToolsReduxState) => ReduxDevToolsReduxState;

  /**
   * Action types to be hidden in the monitors (while passed to the reducers).
   */
  actionsBlacklist?: string | string[];

  /**
   * Action types to be shown in the monitors (while passed to the reducers).
   */
  actionsWhitelist?: string | string[];

  /**
   * Called for every action before sending, takes state and action object, and
   * returns true in case it allows sending the current data to the monitor.
   * Use it as a more advanced version of actionsBlacklist/actionsWhitelist
   * parameters.
   */
  predicate?: (
    state: ReduxDevToolsReduxState,
    action: ReduxDevToolsReduxAction,
  ) => boolean;

  /**
   * If specified as false, it will not record the changes till clicking on
   * Start recording button. Default is true. Available only for Redux
   * enhancer, for others use autoPause.
   */
  shouldRecordChanges?: boolean;

  /**
   * If specified, whenever clicking on Pause recording button and there are
   * actions in the history log, will add this action type. If not specified,
   * will commit when paused. Available only for Redux enhancer.
   * @default @@PAUSED.
   */
  pauseActionType?: string;

  /**
   * Auto pauses when the extension’s window is not opened, and so has zero
   * impact on your app when not in use. Not available for Redux enhancer
   * (as it already does it but storing the data to be sent).
   * @default false
   */
  autoPause?: boolean;

  /**
   * If specified as true, it will not allow any non-monitor actions to be
   * dispatched until clicking on Unlock changes button. Available only for
   * Redux enhancer.
   * @default false
   */
  shouldStartLocked?: boolean;

  /**
   * If set to false, will not recompute the states on hot reloading (or on
   * replacing the reducers). Available only for Redux enhancer.
   * @default true
   */
  shouldHotReload?: boolean;

  /**
   * If specified as true, whenever there's an exception in reducers, the
   * monitors will show the error message, and next actions will not be
   * dispatched.
   */
  shouldCatchErrors?: boolean;

  /**
   * Specify features to enable/disable.
   */
  features?: ReduxDevToolsFeaturesPropertyOptions;
}
