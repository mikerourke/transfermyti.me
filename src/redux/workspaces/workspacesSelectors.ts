import { compose, isNil, prop, sortBy, toLower } from "ramda";

import { selectIdToLinkedId } from "~/api/selectIdToLinkedId";
import { createSelector } from "~/redux/reduxTools";
import type { Mapping, ReduxState, Workspace } from "~/types";

export const areWorkspacesFetchingSelector = (state: ReduxState): boolean =>
  state.workspaces.isFetching;

export const activeWorkspaceIdSelector = (state: ReduxState): string =>
  state.workspaces.activeWorkspaceId;

export const sourceWorkspacesByIdSelector = (
  state: ReduxState,
): Dictionary<Workspace> => state.workspaces.source;

export const targetWorkspacesByIdSelector = (
  state: ReduxState,
): Dictionary<Workspace> => state.workspaces.target;

export const sourceWorkspacesSelector = createSelector(
  sourceWorkspacesByIdSelector,
  (workspacesById): Workspace[] => {
    const workspaces = Object.values(workspacesById);

    return sortBy(compose(toLower, prop("name")), workspaces);
  },
);

export const firstIncludedWorkspaceIdSelector = createSelector(
  sourceWorkspacesSelector,
  (sourceWorkspaces): string => {
    const firstIncluded = sourceWorkspaces.find(
      (sourceWorkspace) => sourceWorkspace.isIncluded,
    );

    if (firstIncluded === undefined) {
      return "";
    }

    return firstIncluded.id;
  },
);

export const includedSourceWorkspacesSelector = createSelector(
  sourceWorkspacesSelector,
  (workspaces): Workspace[] =>
    workspaces.filter((workspace) => workspace.isIncluded),
);

export const sourceIncludedWorkspacesCountSelector = createSelector(
  includedSourceWorkspacesSelector,
  (includedWorkspaces): number => includedWorkspaces.length,
);

export const sourceWorkspacesForTransferSelector = createSelector(
  includedSourceWorkspacesSelector,
  (includedWorkspaces): Workspace[] =>
    includedWorkspaces.filter((workspace) => isNil(workspace.linkedId)),
);

export const workspaceIdToLinkedIdSelector = createSelector(
  sourceWorkspacesByIdSelector,
  (sourceWorkspacesById): Dictionary<string> =>
    selectIdToLinkedId(sourceWorkspacesById),
);

export const targetWorkspacesSelector = createSelector(
  targetWorkspacesByIdSelector,
  (workspacesById): Workspace[] => {
    const workspaces = Object.values(workspacesById);

    return sortBy(compose(toLower, prop("name")), workspaces);
  },
);

export const missingTargetWorkspacesSelector = createSelector(
  includedSourceWorkspacesSelector,
  (includedSourceWorkspaces): Workspace[] =>
    includedSourceWorkspaces.filter(({ linkedId }) => linkedId === null),
);

export const hasDuplicateTargetWorkspacesSelector = createSelector(
  includedSourceWorkspacesSelector,
  (includedSourceWorkspaces): boolean => {
    const countByLinkedId: Dictionary<number> = {};

    for (const { linkedId } of includedSourceWorkspaces) {
      if (linkedId === null) {
        continue;
      }

      if (linkedId in countByLinkedId) {
        countByLinkedId[linkedId] += 1;
      } else {
        countByLinkedId[linkedId] = 1;
      }
    }

    return Object.values(countByLinkedId).some((count) => count > 1);
  },
);

const limitIdsToIncluded = (workspaces: Workspace[]): string[] => {
  const includedIds: string[] = [];

  for (const workspace of workspaces) {
    if (workspace.isIncluded) {
      includedIds.push(workspace.id);
    }
  }

  return includedIds;
};

export const includedSourceWorkspaceIdsSelector = createSelector(
  sourceWorkspacesSelector,
  limitIdsToIncluded,
);

const includedTargetWorkspaceIdsSelector = createSelector(
  targetWorkspacesSelector,
  limitIdsToIncluded,
);

export const includedWorkspaceIdsByMappingSelector = createSelector(
  includedSourceWorkspaceIdsSelector,
  includedTargetWorkspaceIdsSelector,
  (
    includedSourceWorkspaceIds,
    includedTargetWorkspaceIds,
  ): Record<Mapping, string[]> => ({
    source: includedSourceWorkspaceIds,
    target: includedTargetWorkspaceIds,
  }),
);
