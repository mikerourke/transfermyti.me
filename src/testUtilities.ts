export { FAKES } from "~/redux/__fakes__/state";

type MockInstance = jest.MockInstance<AnyValid, AnyValid>;

interface WrapForMockingReturn {
  calls: AnyValid[];
  clear(): MockInstance;
  implementOnce(funcOrReturn: AnyValid): MockInstance;
  rejectAlways(err?: AnyValid): MockInstance;
  rejectOnce(err?: AnyValid): MockInstance;
  resolveMultiple(values: AnyValid[]): void;
  resolveOnce(value?: AnyValid): MockInstance;
  returnOnce(value?: AnyValid): MockInstance;
  returnMultiple(values: AnyValid[]): void;
  throwOnce(message?: string): MockInstance;
}

/**
 * Wrappers around someFunc.mock...() that Jest provides. Using
 * this avoids the annoying `someFunc as AnyValid` required to shut TypeScript
 * up.
 */
export function wrapForMocking(func: AnyValid): WrapForMockingReturn {
  return {
    calls: func?.mock?.calls,
    clear(): MockInstance {
      return func.mockClear();
    },
    implementOnce(funcOrReturn: AnyValid): MockInstance {
      if (typeof funcOrReturn === "function") {
        return func.mockImplementationOnce(funcOrReturn);
      } else {
        return func.mockImplementationOnce(() => funcOrReturn);
      }
    },
    rejectAlways(err: AnyValid = "error"): MockInstance {
      return func.mockRejectedValue(
        typeof err === "string" ? new Error(err) : err,
      );
    },
    rejectOnce(err: AnyValid = "error"): MockInstance {
      return func.mockRejectedValueOnce(
        typeof err === "string" ? new Error(err) : err,
      );
    },
    resolveMultiple(values: AnyValid[]): void {
      for (const value of values) {
        func.mockResolvedValueOnce(value);
      }
    },
    resolveOnce(value: AnyValid = null): MockInstance {
      return func.mockResolvedValueOnce(value);
    },
    returnOnce(value: AnyValid = null): MockInstance {
      return func.mockReturnValueOnce(value);
    },
    returnMultiple(values: AnyValid[]): void {
      for (const value of values) {
        func.mockReturnValueOnce(value);
      }
    },
    throwOnce(message: string = "error"): MockInstance {
      return func.mockImplementationOnce(() => {
        throw new Error(message);
      });
    },
  };
}

/**
 * Allows you to get around `delay()` calls in sagas.
 * @see https://github.com/jfairbank/redux-saga-test-plan/issues/257#issuecomment-503016758
 * @example
 *   return expectSaga(someSaga)
 *     .provide([
 *       { call: provideDelay },
 *     ])
 *     ...
 */
export function provideDelay(
  { fn }: AnyValid,
  next: AnyValid,
): null | VoidFunction {
  return fn.name === "delayP" ? null : next();
}
