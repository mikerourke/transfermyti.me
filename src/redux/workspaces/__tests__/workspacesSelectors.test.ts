import { describe, expect, test } from "vitest";

import { FAKES } from "~/testUtilities";
import { EntityGroup } from "~/types";

import * as workspacesSelectors from "../workspacesSelectors";

const { REDUX_STATE, CLOCKIFY_WORKSPACE } = FAKES;

const MOCK_STATE = {
  ...REDUX_STATE,
  workspaces: {
    ...REDUX_STATE.workspaces,
    source: {
      ...REDUX_STATE.workspaces.source,
      "1002": {
        id: "1002",
        name: "Test Workspace 2",
        userIds: ["6001"],
        isAdmin: true,
        isPaid: false,
        workspaceId: "1002",
        entryCount: 20,
        linkedId: null,
        isIncluded: true,
        memberOf: EntityGroup.Workspaces,
      },
    },
  },
};

describe("within workspacesSelectors", () => {
  describe("selectors that directly access state return the correct value", () => {
    const testCases = [
      {
        name: "areWorkspacesFetchingSelector returns state.isFetching",
        selector: workspacesSelectors.areWorkspacesFetchingSelector,
        expected: MOCK_STATE.workspaces.isFetching,
      },
      {
        name: "activeWorkspaceIdSelector returns values from state.activeWorkspaceId",
        selector: workspacesSelectors.activeWorkspaceIdSelector,
        expected: MOCK_STATE.workspaces.activeWorkspaceId,
      },
      {
        name: "sourceWorkspacesByIdSelector returns state.source",
        selector: workspacesSelectors.sourceWorkspacesByIdSelector,
        expected: MOCK_STATE.workspaces.source,
      },
      {
        name: "sourceWorkspacesSelector returns array of values from state.source",
        selector: workspacesSelectors.sourceWorkspacesSelector,
        expected: Object.values(MOCK_STATE.workspaces.source),
      },
      {
        name: "targetWorkspacesByIdSelector returns state.target",
        selector: workspacesSelectors.targetWorkspacesByIdSelector,
        expected: MOCK_STATE.workspaces.target,
      },
    ];

    for (const testCase of testCases) {
      test.concurrent(testCase.name, async () => {
        const result = testCase.selector(MOCK_STATE);

        expect(result).toEqual(testCase.expected);
      });
    }
  });

  describe("the selectors match their snapshots", () => {
    const testCases = [
      {
        name: "for the includedSourceWorkspacesSelector",
        selector: workspacesSelectors.includedSourceWorkspacesSelector,
      },
      {
        name: "for the sourceWorkspacesForTransferSelector",
        selector: workspacesSelectors.sourceWorkspacesForTransferSelector,
      },
      {
        name: "for the workspaceIdToLinkedIdSelector",
        selector: workspacesSelectors.workspaceIdToLinkedIdSelector,
      },
      {
        name: "for the missingTargetWorkspacesSelector",
        selector: workspacesSelectors.missingTargetWorkspacesSelector,
      },
      {
        name: "for the includedSourceWorkspaceIdsSelector",
        selector: workspacesSelectors.includedSourceWorkspaceIdsSelector,
      },
      {
        name: "for the includedWorkspaceIdsByMappingSelector",
        selector: workspacesSelectors.includedWorkspaceIdsByMappingSelector,
      },
    ];

    for (const testCase of testCases) {
      test.concurrent(testCase.name, async () => {
        const result = testCase.selector(MOCK_STATE);

        expect(result).toMatchSnapshot();
      });
    }
  });

  describe("the firstIncludedWorkspaceIdSelector", () => {
    test("returns the ID of the first included workspace if one is present in state", () => {
      const result = workspacesSelectors.firstIncludedWorkspaceIdSelector(MOCK_STATE);

      expect(result).toBe("1001");
    });

    test("returns an empty string if no included workspaces are present", () => {
      const updatedState = {
        ...MOCK_STATE,
        workspaces: {
          ...MOCK_STATE.workspaces,
          source: {
            "1001": {
              // @ts-expect-error
              ...MOCK_STATE.workspaces.source["1001"],
              isIncluded: false,
            },
          },
          target: {
            [CLOCKIFY_WORKSPACE.id]: {
              ...CLOCKIFY_WORKSPACE,
            },
          },
        },
      };

      const result = workspacesSelectors.firstIncludedWorkspaceIdSelector(updatedState);

      expect(result).toBe("");
    });
  });

  describe("the hasDuplicateTargetWorkspacesSelector", () => {
    test("returns false if not duplicate linkedIds are present", () => {
      const result = workspacesSelectors.hasDuplicateTargetWorkspacesSelector(MOCK_STATE);

      expect(result).toBe(false);
    });

    test("returns true if duplicate linkedIds are present", () => {
      const updatedState = {
        ...MOCK_STATE,
        workspaces: {
          ...MOCK_STATE.workspaces,
          source: {
            ...MOCK_STATE.workspaces.source,
            "1001": {
              // @ts-expect-error
              ...MOCK_STATE.workspaces.source["1001"],
              linkedId: "clock-workspace-01",
            },
            "1002": {
              ...MOCK_STATE.workspaces.source["1002"],
              linkedId: "clock-workspace-01",
            },
          },
        },
      };

      const result = workspacesSelectors.hasDuplicateTargetWorkspacesSelector(updatedState);

      expect(result).toBe(true);
    });
  });

  test("the includedWorkspaceIdsByMappingSelector ignores workspaces that aren't included", () => {
    const updatedState = {
      ...MOCK_STATE,
      workspaces: {
        ...MOCK_STATE.workspaces,
        source: {
          ...MOCK_STATE.workspaces.source,
          "1002": {
            ...MOCK_STATE.workspaces.source["1002"],
            isIncluded: false,
          },
        },
        target: {
          [CLOCKIFY_WORKSPACE.id]: {
            ...CLOCKIFY_WORKSPACE,
          },
        },
      },
    };

    const result = workspacesSelectors.includedWorkspaceIdsByMappingSelector(updatedState);

    expect(result).toEqual({
      source: ["1001"],
      target: [CLOCKIFY_WORKSPACE.id],
    });
  });

  test("the sourceIncludedWorkspacesCountSelector returns the count of included source workspaces", () => {
    const result = workspacesSelectors.sourceIncludedWorkspacesCountSelector(MOCK_STATE);

    expect(result).toBe(2);
  });
});
