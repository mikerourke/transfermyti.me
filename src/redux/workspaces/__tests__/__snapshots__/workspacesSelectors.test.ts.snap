// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`within workspacesSelectors > the selectors match their snapshots > for the includedSourceWorkspaceIdsSelector 1`] = `
[
  {
    "entryCount": 20,
    "id": "1001",
    "isAdmin": true,
    "isIncluded": true,
    "isPaid": false,
    "linkedId": "clock-workspace-01",
    "memberOf": "workspaces",
    "name": "Test Workspace",
    "userIds": [
      "6001",
    ],
    "workspaceId": "1001",
  },
  {
    "entryCount": 20,
    "id": "1002",
    "isAdmin": true,
    "isIncluded": true,
    "isPaid": false,
    "linkedId": null,
    "memberOf": "workspaces",
    "name": "Test Workspace 2",
    "userIds": [
      "6001",
    ],
    "workspaceId": "1002",
  },
]
`;

exports[`within workspacesSelectors > the selectors match their snapshots > for the includedSourceWorkspaceIdsSelector 2`] = `
[
  {
    "entryCount": 20,
    "id": "1002",
    "isAdmin": true,
    "isIncluded": true,
    "isPaid": false,
    "linkedId": null,
    "memberOf": "workspaces",
    "name": "Test Workspace 2",
    "userIds": [
      "6001",
    ],
    "workspaceId": "1002",
  },
]
`;

exports[`within workspacesSelectors > the selectors match their snapshots > for the includedSourceWorkspaceIdsSelector 3`] = `
{
  "1001": "clock-workspace-01",
  "clock-workspace-01": "1001",
}
`;

exports[`within workspacesSelectors > the selectors match their snapshots > for the includedSourceWorkspaceIdsSelector 4`] = `
[
  {
    "entryCount": 20,
    "id": "1002",
    "isAdmin": true,
    "isIncluded": true,
    "isPaid": false,
    "linkedId": null,
    "memberOf": "workspaces",
    "name": "Test Workspace 2",
    "userIds": [
      "6001",
    ],
    "workspaceId": "1002",
  },
]
`;

exports[`within workspacesSelectors > the selectors match their snapshots > for the includedSourceWorkspaceIdsSelector 5`] = `
[
  "1001",
  "1002",
]
`;

exports[`within workspacesSelectors > the selectors match their snapshots > for the includedWorkspaceIdsByMappingSelector 1`] = `
{
  "source": [
    "1001",
    "1002",
  ],
  "target": [
    "clock-workspace-01",
  ],
}
`;
